plugin/src/main/java/io/jenkins/plugins/casc/SecretSourceResolver.java

```diff
@@ -19,64 +23,117 @@
     private static final String enclosedBy = "${";
     private static final String enclosedIn = "}";
     private static final char escapedWith = '^';
-    private static final String defaultDelimiter = ":-";
+    private static final String escapeEnclosedBy = escapedWith + enclosedBy;
 
     private static final Logger LOGGER = Logger.getLogger(SecretSourceResolver.class.getName());
+    private final StringSubstitutor nullSubstitutor;
+    private final StringSubstitutor substitutor;
+
+    public SecretSourceResolver(ConfigurationContext configurationContext) {
+        ConfigurationContext context = configurationContext;
+        substitutor = new StringSubstitutor(
+            StringLookupFactory.INSTANCE.interpolatorStringLookup(ImmutableMap
+                    .of("base64", StringLookupFactory.INSTANCE.base64EncoderStringLookup(),
+                        "file", FileStringLookup.INSTANCE
+                    ),
+                new ConfigurationContextStringLookup(context), false))
+            .setEscapeChar(escapedWith)
+            .setVariablePrefix(enclosedBy)
+            .setVariableSuffix(enclosedIn)
+            .setEnableSubstitutionInVariables(true)
+            .setPreserveEscapes(true);
+        nullSubstitutor = new StringSubstitutor(UnresolvedLookup.INSTANCE)
+            .setEscapeChar(escapedWith)
+            .setVariablePrefix(enclosedBy)
+            .setVariableSuffix(enclosedIn);
+    }
 
     /**
      * Encodes String so that it can be safely represented in the YAML after export.
      * @param toEncode String to encode
      * @return Encoded string
      * @since 1.25
      */
-    public static String encode(@CheckForNull String toEncode) {
+    public String encode(@CheckForNull String toEncode) {
         if (toEncode == null) {
             return null;
         }
-        return toEncode.replace("${", "^${");
+        return toEncode.replace(enclosedBy, escapeEnclosedBy);
     }
 
+    /**
+     * Resolve string with potential secrets
+     *
+     * @param context Configuration context
+     * @param toInterpolate potential variables that need to revealed
+     * @return original string with any secrets that could be resolved if secrets could not be resolved they will be defaulted to default value defined by ':-', otherwise default to empty String secrets are defined as anything enclosed by '${}'
+     * @deprecated use ${link {@link #resolve(String)}} instead.
+     * @since 1.42
+     */
+    @Deprecated
     public static String resolve(ConfigurationContext context, String toInterpolate) {
-        return substitutor(context).replace(toInterpolate);
+        return context.getSecretSourceResolver().resolve(toInterpolate);
     }
```

Comment:
Could potentially remove this as usually this is only called directly from test packages in other plugins ie. HashiCorp Vault plugin.
 
Depeneds if we are okay with yet another breakage 😆